content/post/review-building-static-analysis-tools.md

```diff
@@ -0,0 +1,249 @@
+
+## Source code analysis recap
+
+Before digging deeper into the paper, a bit of the context on what kind of analysis in general is applicable to programs. There are two main type of program analysis:
+
+* *Static code analysis*, looks only at the source code, without running the program
+* *Dynamic code analysis*, when (potentially customized) program is executed and results are analyzed
+
+This paper talks only about *static code analysis*, which is traditionally further subdivided into two main flavors:
+
+* *intra-procedural,* when analysis is localized inside a small part of the program like a function/class/file
+* *inter-procedural,* that uses broader, whole-program artefacts as input, and conduct more sophisticated analysis (call graph, formal logic, context-sensitive analysis, etc)
+
+According to the paper, Google only runs simpler or *intra-procedural* type of analysis — the only one feasible to conduct at scale of 2 bln LoC.
+
+Interesting enough, this is somehow different from Facebook’s project [Infer](http://fbinfer.com/) approach, which developed a way scale a particular type of [more complex](https://code.facebook.com/posts/1537144479682247/finding-inter-procedural-bugs-at-scale-with-infer-static-analyzer/), *inter-procedural* [separation logic-based](https://en.wikipedia.org/wiki/Separation_logic) compositional analysis, that paper authors are well aware of.
+
+[**Infer static analysis tool**](http://fbinfer.com/)
+TK does hugo have nice big links?
+
+The reasoning behind why Google is not interested in more complex analysis at this point are explicitly listed in the paper:
+
+* *Large investment.* Although theoretically better and more complex analysis exist, it will require non-trivial engineering effort to scale
+
+* *High upfront cost.* It’s very hard to get the data, even to just estimate the cost-benefit ratio for such improvements, to justify research/implementation efforts
+
+* *Work still would be needed to reduce false-positive rates*
+
+* *Finding bugs is easy.* When a codebase is large enough, it will contain practically any imaginable code patter. Many more low-hanging fruits exist using simpler approaches
+
+I find it to be particularly curious, as none of these conditions are set in stone:
+
+* data is hard to get without decrease in productivity of engineers inside the company, due to the way false-positives are measured (more on it in the next section)
```

Comment:
without decreas**ing** ~in~ the productivity

Reply:
![screen shot 2018-08-08 at 17 25 45](https://user-images.githubusercontent.com/2793551/43847101-24561a7e-9b30-11e8-93c0-ae85bf0b3844.png)
![screen shot 2018-08-08 at 17 25 55](https://user-images.githubusercontent.com/2793551/43847102-24702900-9b30-11e8-9d50-eb37860316ef.png)

Reply:
<img width="286" alt="screen shot 2018-08-08 at 5 51 18 pm" src="https://user-images.githubusercontent.com/5582506/43848657-b8843f48-9b33-11e8-8a58-d81f8d89b81e.png">

Reply:
LOL @bzz 
![screen shot 2018-08-08 at 17 59 32](https://user-images.githubusercontent.com/2793551/43849071-d89cf5c6-9b34-11e8-8749-ad60ac359c3a.png)